core/src/main/scala/kafka/log/remote/RemoteStorageManager.scala

```diff
@@ -0,0 +1,138 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.log.remote
+
+import java._
+import java.io.IOException
+
+import kafka.log.LogSegment
+import org.apache.kafka.common.annotation.InterfaceStability
+import org.apache.kafka.common.record.Records
+import org.apache.kafka.common.{Configurable, TopicPartition}
+
+/**
+ * RemoteStorageManager is an interface that allows to plugin different remote storage implementations to copy and
+ * retrieve the log segments.
+ *
+ * All these APIs are still experimental.
+ */
+@InterfaceStability.Unstable
+trait RemoteStorageManager extends Configurable with AutoCloseable {
```

Comment:
@Hangleton In case you're interested, there's a [proposed `S3RemoteStorageManager` implementation](https://github.com/harshach/kafka/pull/18/files) that's being reviewed. Currently, it doesn't do file overwrites on S3, so the only state that is eventually being reached is file existence. This can be solved with some sensible retry policy.

We'll have to invent some consistency mechanism if overwriting starts happening at some point.

Reply:
@Hangleton @ivanyu RLM expects remote logs to be available once they are copied successfully. But it looks like S3 gives eventually consistency model which may cause reads to fail as that object may not yet be available. Are there any guarantees that the uploaded object will be available after a certain duration? As the committed log segments are copied as soon as they are available, this issue may not arise in normal scenarios but it is better to be discussed and set the right contracts with the API.  

There can be other issues like create a topic again after a delete but earlier remote log segments in S3 may still be available which can cause issues. [KIP-516](https://cwiki.apache.org/confluence/display/KAFKA/KIP-516%3A+Topic+Identifiers) gives topic-id as identifier which resolves uniqueness of a topic but that is not yet available. This may be helpful for create after delete scenario but it is not yet available in Kafka.

Did you think of addressing these kind of issues by maintaining remote log metadata in a key/value store with strong consistency or do you want to mitigate these kind of issues with a retry policy?

Reply:
> Are there any guarantees that the uploaded object will be available after a certain duration?

Unfortunately, it's unbounded.

Reply:
Yes the time for convergence is indeterminate and we cannot rely on retries for that.
The consistency model for S3 is

- Strongly consistent on PUT only if no HEAD or GET requests were issued on the key name before creating the object.
- Eventually consistent for PUT overwriting an existing object.
- Eventually consistent for LIST after a PUT.
- Eventually consistent for LIST and GET after a DELETE object.

In this case if the `RemoteStorageManager` requires strong consistency after a mutating call, an additional index would need to be used.